[Durable Objects] Explain blockConcurrencyWhile() (#12467)
cloudflare/cloudflare-docs: content/durable-objects/reference/in-memory-state.md

However, in applications with more complex state, explicitly storing state in your Object may be easier than making transactional storage API calls on every access. Depending on the configuration of your project, write your code in the way that is easiest for you.

{{</Aside>}}

##  `state.blockConcurrencyWhile()` method

> `state.blockConcurrencyWhile()` is a method that allows you to run critical sections of the code while guaranteeing ordering.

Reviewer: Maybe add something along:
… guaranteeing ordering, blocking concurrency within an object while the critical section is running.

`state.blockConcurrencyWhile()` executes a callback (which may be `async`) while blocking any other events from being delivered to the object until the callback completes. 

This allows you to execute code that performs I/O (such as a `fetch()`) with the guarantee that the object's state will not unexpectedly change as a result of concurrent events. All events that were not explicitly initiated as part of the callback itself will be blocked. 

Events that will be blocked include new incoming requests and responses to outgoing requests (such as `fetch()`) that were initiated outside of the callback. Once the callback completes, these events will be delivered.

> `state.blockConcurrencyWhile()` is useful within the constructor of your object to perform initialization that must occur before any requests are delivered.

Reviewer: Although it's useful in the constructor, it may be useful anywhere else, if you want to guarantee that some critical section will complete without any concurrent changes to the object. E.g. you want to run a sequence of storage operations and don't want to run into the risk of some concurrent action changing the status of the storage. In this case you'd run your sequence of operations inside the `blockConcurrencyWhile` and assure that nothing interfered.

If the callback throws an exception, the object will be terminated and reset. This ensures that the object cannot be left stuck in an uninitialized state if something fails unexpectedly. To avoid this behavior, enclose the body of your callback in a `try...catch` block to ensure it cannot throw an exception.

`state.blockConcurrencyWhile()` takes effects right away, pausing everything else, other than the currently executing event and any I/O kicked off from within the `state.blockConcurrencyWhile()` callback. The value returned by the callback becomes the value return by `state.blockConcurrencyWhile()` itself. 

> To simulate eviction, use `state.blockConcurrencyWhile()`. Inside the callback, throw an exception. Throwing an exception causes the system to recreate the in-memory Object, without affecting the durable storage. Alternatively, use the `abort()` method. 

Reviewer: Uncaught exceptions inside the `blockConcurrencyWhile()` will break the actor.

However, in applications with more complex state, explicitly storing state in your Object may be easier than making transactional storage API calls on every access. Depending on the configuration of your project, write your code in the way that is easiest for you.

{{</Aside>}}

##  `state.blockConcurrencyWhile()` method

`state.blockConcurrencyWhile()` is a method that guarantees ordering, blocking concurrency within an object while the critical section is running.

> `state.blockConcurrencyWhile()` executes a callback (which may be `async`) while blocking any other events from being delivered to the object until the callback completes. 

Reviewer 2: Let's change this to:

"`state.blockConcurrencyWhile(callback) -> Promise<T>` executes a callback (which may be `async`) with return type `T`, while blocking any other events from being delivered to the object until the callback completes."

> `state.blockConcurrencyWhile(callback)` has a return type of `Promise<T>`, where `T` is the return type of `callback`. `state.blockConcurrencyWhile()` executes a callback (which may be `async`) while blocking any other events from being delivered to the object until the callback completes

Reviewer 3: ```suggestion
`state.blockConcurrencyWhile(callback)` has a return type of `Promise<T>`, where `T` is the return type of `callback`. `state.blockConcurrencyWhile()` executes a callback (which may be `async`) while blocking any other events from being delivered to the object until the callback completes.
```

> `state.blockConcurrencyWhile(callback)` has a return type of `Promise<T>`, where `T` is the return type of callback. `state.blockConcurrencyWhile()` executes a callback (which may be `async`) with return type T, while blocking any other events from being delivered to the object until the callback completes

Reviewer 2: Can the "callback" at the end of the first sentence (line 51) be `callback` (code block), since it's an argument?

Also, can we shorten the second sentence to:
- `state.blockConcurrencyWhile()` executes the callback (which may be `async`) while blocking...

We're currently restating "return type `T`" which feels a bit clunky.

This allows you to execute code that performs I/O (such as a `fetch()`) with the guarantee that the object's state will not unexpectedly change as a result of concurrent events. All events that were not explicitly initiated as part of the callback itself will be blocked. 

Events that will be blocked include new incoming requests and responses to outgoing requests (such as `fetch()`) that were initiated outside of the callback. Once the callback completes, these events will be delivered.

> `state.blockConcurrencyWhile()` is useful within the constructor of your object to perform initialization that must occur before any requests are delivered. `state.blockConcurrencyWhile()` may also be useful anywhere else other than within the constructor. For example, if you want to run a sequence of storage operations and want to avoid concurrent actions changing the status of the storage, run your sequence of operations inside the `state.blockConcurrencyWhile()` and no operation will be interfered.

Reviewer 2: nit: At the end, change:

- "inside the `state.blockConcurrencyWhile()` and no operation will be interfered." to
- "inside of a `state.blockConcurrencyWhile()` callback.".

> `state.blockConcurrencyWhile()` is useful within the constructor of your object to perform initialization that must occur before any requests are delivered. `state.blockConcurrencyWhile()` may also be useful outside of the constructor. For example, if you want to run a sequence of storage operations and want to avoid concurrent actions changing the status of the storage, run your sequence of operations inside the `state.blockConcurrencyWhile()` callback.

Reviewer 4: ```suggestion
`state.blockConcurrencyWhile()` is useful within the constructor of your object to perform initialization that must occur before any requests are delivered. This method may also be useful outside of the constructor. For example, if you want to run a sequence of storage operations and want to avoid concurrent actions changing the status of the storage, run your sequence of operations inside the `state.blockConcurrencyWhile()` callback.
```

> `state.blockConcurrencyWhile()` is useful within the constructor of your object to perform initialization that must occur before any requests are delivered. `state.blockConcurrencyWhile()` may also be useful anywhere else other than within the constructor. For example, if you want to run a sequence of storage operations and want to avoid concurrent actions changing the status of the storage, run your sequence of operations inside the `state.blockConcurrencyWhile()` callback.

Reviewer 2: **nit**: "..._may also be useful anywhere else other than within the constructor_." feels a bit confusing to me. It almost makes it feel like we're saying "it's not useful within the constructor".

Could we change this to
- "..._may also be useful outside of the constructor_."

Reviewer 2: Ah, and on the same line (a sentence later) "_changing the status of the storage_" would be better as "_changing the state of storage_".

If the callback throws an exception, the object will be terminated and reset. This ensures that the object cannot be left stuck in an uninitialized state if something fails unexpectedly. To avoid this behavior, enclose the body of your callback in a `try...catch` block to ensure it cannot throw an exception.

> `state.blockConcurrencyWhile()` takes effects right away, pausing everything else, other than the currently executing event and any I/O kicked off from within the `state.blockConcurrencyWhile()` callback. The value returned by the callback becomes the value returned by `state.blockConcurrencyWhile()` itself. 

Reviewer 2: `The value returned by the callback becomes the value returned by `state.blockConcurrencyWhile()` itself.`  It's nice to see this written out, but I think we should also explicitly say that `blockConcurrencyWhile()` has a return type of `Promise<T>`.

Let's say up on line 49 we do something like `state.blockConcurrencyWhile(callback) -> Promise<T>`, where `T` is the return type of `callback`.

> `state.blockConcurrencyWhile()` takes effects right away, pausing everything else, other than the currently executing event and any I/O kicked off from within the `state.blockConcurrencyWhile()` callback. The value returned by the callback becomes the value returned by `state.blockConcurrencyWhile()` itself. 

Reviewer 4: ```suggestion
`state.blockConcurrencyWhile()` takes effect right away, pausing everything else, other than the currently executing event and any I/O started from within the `state.blockConcurrencyWhile()` callback. The value returned by the callback becomes the value returned by `state.blockConcurrencyWhile()` itself. 
```

> To simulate eviction, use `state.blockConcurrencyWhile()`. Inside the callback, throw an exception. Throwing an exception causes the system to recreate the in-memory object, without affecting the durable storage. Uncaught exceptions inside the `state.blockConcurrencyWhile()` will break the actor. Alternatively, use the `abort()` method to simulate eviction. 

Reviewer 2: We need to delete this last sentence: "Alternatively, use the `abort()` method to simulate eviction."

This is an experimental feature and my understanding is we have do not intend on supporting it externally.
